react/features/base/settings/functions.js

```diff
@@ -5,6 +5,119 @@ import { toState } from '../redux';
 
 import { DEFAULT_SERVER_URL } from './constants';
 
+/**
+ * Searches known devices for a matching deviceId and fall back to matching on
+ * label. Returns the stored preferred cameraDeviceId if a match is not found.
+ *
+ * @param {Object|Function} stateful - The redux state object or
+ * {@code getState} function.
+ * @returns {string}
+ */
+export function getPreferredCameraDeviceId(stateful: Object | Function) {
+    const state = toState(stateful);
+    const {
+        cameraDeviceId,
+        cameraDeviceLabel
+    } = state['features/base/settings'];
+
+    // If there is no label at all, there is no need to fall back to checking
+    // the label for a fuzzy match. And then if there is no cameraDeviceId,
+    // there is no searching to perform.
+    if (!cameraDeviceLabel || !cameraDeviceId) {
+        return cameraDeviceId;
+    }
+
+    const { videoInput } = state['features/base/devices'];
+    const foundMatchingBasedonDeviceId = videoInput.find(
+        candidate => candidate.deviceId === cameraDeviceId);
+
+    // Prioritize matching the deviceId of the camera.
+    if (foundMatchingBasedonDeviceId) {
+        return cameraDeviceId;
+    }
+
+    // Fall back to doing a partial match of the label if a matching deviceId
+    // has not been found. Operating systems may append " #{number}" somewhere
+    // in the label so find and strip that bit.
+    const lastSpacePoundNumRegex = /\s#\d*(?!.*\s#\d*)/;
+    const replacement = '';
+    const preferredDeviceLabel
+        = cameraDeviceLabel.replace(lastSpacePoundNumRegex, replacement);
+    const foundMatchBasedOnLabel = videoInput.find(candidate => {
+        const { label } = candidate;
+
+        if (!label) {
+            return false;
+        } else if (cameraDeviceLabel === label) {
+            return true;
+        }
+
+        const candidateLabel
+            = label.replace(lastSpacePoundNumRegex, replacement);
+
+        return preferredDeviceLabel === candidateLabel;
+    });
+
+    return foundMatchBasedOnLabel
+        ? foundMatchBasedOnLabel.deviceId : cameraDeviceId;
+}
+
+/**
+ * Searches known devices for a matching deviceId and fall back to matching on
+ * label. Returns the stored preferred micDeviceId if a match is not found.
+ *
+ * @param {Object|Function} stateful - The redux state object or
+ * {@code getState} function.
+ * @returns {string}
+ */
+export function getPreferredMicDeviceId(stateful: Object | Function) {
```

Comment:
is there not a way to have a helper function which does the common logic (label fallback, stripping of the os suffix, etc.) that both audio and video device logic could use?

Reply:
I couldn't think of it without passing in several options, at which point I wondered if what worthwhile. For example, I look at different keys in features/base/settings, I look at a different lists of devices in features/base/device, I use a different regex,  and I use a different replacement character. I thought of breaking it up into little helpers too but thought it didn't really make things better. I can keep thinking about it now that I suspect device issues were due to lack of persistence and not the label changing.

Reply:
I've now gone with the approach of passing in options.

Reply:
For the first time I can remember I have blown away changes I wanted to keep by doing a force push; force pushing has finally bit me hard. I redid the changes to abstract the logic into one place.